File: src/c4bb/sim/loader.js
// loader.js - .c4r image loader for the c4bb machine.
//
// A direct port of c4l.c (the cleanest loader in the repo) for 32-bit
// images and a flat arena: code and data are placed at a base address,
// and the patch table rebases every address unambiguously, which is
// what makes loading into an arena trivial.
//
// The load sequence mirrors what native c4m + load-c4r.c do:
//   fw.c4r first (its constructor fills the firmware vector latches),
//   then the program, then all constructors, then main via the boot
//   frame; destructors run only if main returns normally (exit()
//   halts the machine through the POWER device and skips them).

import { MEM_BASE, CONS_RET, OPNAMES_ROM, TLEV_ADDR } from './arena.js';
import { HEAP_BASE, HEAP_END, OPNAMES } from './devices.js';
import { R, OP } from './machine.js';
import { FETCH } from './ucode.js';

const PATCH_CODE = -1, PATCH_DATA = -2, PATCH_DCODE = -3, PATCH_DDATA = -4;

export function parseC4r(bytes) {
  if (bytes.length < 13 || bytes[0] !== 0x43 || bytes[1] !== 0x34 || bytes[2] !== 0x52)
    throw new Error('not a .c4r image');
  const version = bytes[3], wordBits = bytes[4];
  if (wordBits !== 32) throw new Error(`c4bb is a 32-bit machine; image is ${wordBits}-bit`);
  const dv = new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength);
  let p = 13;
  const w = () => { const v = dv.getInt32(p, true); p += 4; return v; };
  const entry = w(), codeLen = w(), dataLen = w(), patchLen = w(),
        symbolsLen = w(), consLen = w(), desLen = w();
  // Format v3 puts the data segment's in-memory size in the word at
  // byte 5, where v2 wrote filler (c4r.lisp, load-c4r.c:64). Anything
  // past DATALEN is BSS: uninitialized globals that occupy no image
  // bytes and must be zero at entry. Ignoring it is not merely a
  // missing feature -- the next thing placed in the arena (the heap,
  // or another image) lands ON TOP of those globals, so a c4lc-built
  // program's uninitialized globals get overwritten by its own
  // malloc'd memory. The arena is already zero-filled, so reserving
  // the space is all that is needed.
  const memsz = Math.max(dataLen, version >= 3 ? dv.getInt32(5, true) : dataLen);

  w();                                       // 'C' marker
  const code = new Int32Array(codeLen);
  for (let i = 0; i < codeLen; i++) code[i] = w();

  w();                                       // 'D' marker
  const data = bytes.slice(p, p + dataLen); p += dataLen;

  w();                                       // 'P' marker
  const patches = [];
  for (let i = 0; i < patchLen; i++) patches.push({ type: w(), addr: w(), val: w() });

  w();                                       // 'c' marker
  const cons = []; for (let i = 0; i < consLen; i++) cons.push(w());
  w();                                       // 'd' marker
  const des = []; for (let i = 0; i < desLen; i++) des.push(w());
  // 'S' symbols follow; not needed to run

  return { version, entry, code, data, memsz, patches, cons, des, symbolsLen };
}

// Place an image in the arena. codeBase word-aligned; data follows.
// Returns { entryAddr, codeBase, dataBase, top, cons, des } with all
// addresses in bytes.
export function loadImage(arena, img, base) {
  const codeBase = (base + 3) & ~3;
  const dataBase = (codeBase + img.code.length * 4 + 7) & ~3;
  for (let i = 0; i < img.code.length; i++)
    arena.i32[(codeBase >> 2) + i] = img.code[i];
  arena.writeBytes(dataBase, img.data);

  for (const pt of img.patches) {
    if (pt.type === PATCH_CODE)
      arena.write32(codeBase + pt.addr * 4, codeBase + pt.val * 4);
    else if (pt.type === PATCH_DATA)
      arena.write32(codeBase + pt.addr * 4, dataBase + pt.val);
    else if (pt.type === PATCH_DCODE)
      arena.write32(dataBase + pt.addr, codeBase + pt.val * 4);
    else if (pt.type === PATCH_DDATA)
      arena.write32(dataBase + pt.addr, dataBase + pt.val);
    // positive types: linker symbol references, skipped (c4l.c:164)
  }

  return {
    entryAddr: codeBase + img.entry * 4,
    codeBase, dataBase,
    top: (dataBase + Math.max(img.data.length, img.memsz | 0) + 7) & ~3,
    cons: img.cons.map(o => codeBase + o * 4),
    des: img.des.map(o => codeBase + o * 4),
  };
}

// Write the fixed low-memory ROM contents.
export function initRom(arena) {
  arena.write32(TLEV_ADDR, OP.TLEV);
  const enc = new TextEncoder().encode(OPNAMES);
  arena.writeBytes(OPNAMES_ROM, enc);
}

// Call a loaded function synchronously: fabricate the frame JSR would
// have made (args pushed left to right, return pc = CONS_RET) and run
// the machine until it comes back. This is the JS-side equivalent of
// c4l.c's invoke stub.
export function callFunction(machine, addr, args, maxCycles = 2e9) {
  const arena = machine.arena, regs = machine.regs;
  let sp = regs[R.SP];
  for (const a of args) { sp -= 4; arena.write32(sp, a | 0); }
  sp -= 4; arena.write32(sp, CONS_RET);
  regs[R.SP] = sp;
  regs[R.PC] = addr;
  machine.upc = FETCH;
  if (!machine.runUntilPc(CONS_RET, maxCycles))
    throw new Error('c4bb: function call did not return');
  regs[R.SP] = (regs[R.SP] + 4 * args.length) | 0;   // caller cleanup (ADJ)
  return regs[R.A];
}

// Full boot: firmware + program + argv + constructors. Afterwards the
// machine is ready to run; when PC reaches CONS_RET, main returned
// (run destructors, status in A). The POWER device handles exit().
// opts.mbox: bytes to carve out below the stack reserve for the mailbox
// (devices.js MBOX_*); the firmware sees a heap that stops short of it,
// so nothing the guest allocates can land on the rings. Zero = not fitted.
// opts.ctorBudget: cycles ONE constructor may take before boot gives up. The
// default is effectively no limit, which is right for an image the machine can
// read; an image it cannot -- a corrupt one, or one built for a different opcode
// ROM -- has no constructors, only whatever the garbage does, and that does not
// return. Lower it when running an image that may be garbage is the point.
export function boot(machine, fwBytes, progBytes, argv, opts = {}) {
  const arena = machine.arena;
  const mboxLen = (opts.mbox | 0) > 0 ? ((opts.mbox | 0) + 4095) & ~4095 : 0;
  initRom(arena);

  const fwImg = loadImage(arena, parseC4r(fwBytes), MEM_BASE);
  // No program image: the FIRMWARE is the program. That is the machine
  // switched on with nothing in ROM but its BIOS, which then goes
  // looking for a medium to boot -- docs/c4bb-storage.md M10.
  const progImg = progBytes ? loadImage(arena, parseC4r(progBytes), fwImg.top + 16)
                            : fwImg;

  // argv strings + pointer table in the reserved area above the stack
  let argPtr = arena.stackTop;
  const argvAddrs = [];
  const enc = new TextEncoder();
  for (const s of argv) {
    const bytes = enc.encode(s);
    arena.writeBytes(argPtr, bytes);
    arena.write8(argPtr + bytes.length, 0);
    argvAddrs.push(argPtr);
    argPtr += bytes.length + 1;
  }
  argPtr = (argPtr + 3) & ~3;
  const argvBase = argPtr;
  for (const a of argvAddrs) { arena.write32(argPtr, a); argPtr += 4; }

  // heap bounds for the firmware allocator (1 MB stack reserve)
  // (opts.reserve: host-owned bytes below the mailbox, outside the heap --
  // where a host parks something the guest must see but never allocate over)
  const reserveLen = (opts.reserve | 0) > 0 ? ((opts.reserve | 0) + 4095) & ~4095 : 0;
  const heapBase = (progImg.top + 4096) & ~4095;
  const heapEnd = ((arena.stackTop - 1024 * 1024) & ~4095) - mboxLen - reserveLen;
  machine.dev.write32(HEAP_BASE, heapBase);
  machine.dev.write32(HEAP_END, heapEnd);
  const reserve = reserveLen ? { base: heapEnd, len: reserveLen } : null;
  const mbox = mboxLen ? { base: heapEnd + reserveLen, len: mboxLen } : null;
  if (mbox) { machine.dev.mbox = mbox; arena.u8.fill(0, mbox.base, mbox.base + mbox.len); }

  // initial stack
  machine.regs[R.SP] = machine.regs[R.BP] = arena.stackTop;

  // constructors: firmware first (fills the vector latches), then the
  // program's, each called with a single argument (c4l.c:188 passes 0)
  const ctorBudget = opts.ctorBudget ?? 2e9;
  for (const c of fwImg.cons) callFunction(machine, c, [0], ctorBudget);
  if (progImg !== fwImg) for (const c of progImg.cons) callFunction(machine, c, [0], ctorBudget);

  // boot frame for main(argc, argv), pushed like c4m.c:1466 but with
  // the CONS_RET sentinel so destructors can run on normal return
  let sp = machine.regs[R.SP];
  sp -= 4; arena.write32(sp, argv.length);        // argc
  sp -= 4; arena.write32(sp, argvBase);           // argv
  sp -= 4; arena.write32(sp, CONS_RET);           // return pc sentinel
  machine.regs[R.SP] = sp;
  machine.regs[R.PC] = progImg.entryAddr;
  machine.upc = FETCH;

  return { fwImg, progImg, argvBase, heapBase, heapEnd, mbox, reserve };
}

// Run to completion after boot(); returns the exit status. When a
// Turbo engine is supplied the main body runs compiled; constructors,
// destructors and the sentinel handling are identical either way.
export function runToExit(machine, progImg, maxCycles = Infinity, turbo = null) {
  const limit = machine.cycle + maxCycles;
  while (!machine.dev.halted && machine.cycle < limit) {
    if (machine.upc === FETCH && machine.regs[R.PC] === CONS_RET) {
      // main returned normally: run destructors, then done
      const status = machine.regs[R.A];
      machine.regs[R.SP] = (machine.regs[R.SP] + 8) | 0;  // pop argc/argv
      for (const d of progImg.des) callFunction(machine, d, []);
      return status;
    }
    if (turbo) {
      turbo.run(limit - machine.cycle, CONS_RET);
      if (!machine.dev.halted && machine.regs[R.PC] !== CONS_RET) break;
    } else if (!machine.step()) break;
  }
  if (machine.dev.halted) return machine.dev.status;
  if (machine.regs[R.PC] === CONS_RET) {
    const status = machine.regs[R.A];
    machine.regs[R.SP] = (machine.regs[R.SP] + 8) | 0;
    for (const d of progImg.des) callFunction(machine, d, []);
    return status;
  }
  throw new Error('c4bb: cycle budget exhausted');
}

File: src/c4mp/loader.c
//
// c4mp .c4r loader: read an image, relocate it, hand back an entry
// address. A close port of src/c4ix/loader.c, which is the compact
// modern reader (load-c4r.c is the same job in 1154 lines, written
// against a much older format revision).
//
// The difference from C4IX is what happens next. C4IX resolves the
// constructor and destructor lists and hands them to a task shim,
// because it is a kernel and the shim is guest code. c4mp is the
// machine, so it cannot call guest code at all -- main.c builds a
// bootstrap trampoline out of real instructions instead, and these
// lists become JSR targets in it.
//

#include "c4mp.h"

enum { C4R_BUF_MAX = 4194304 };   // 4MB image limit

static int loader_word(char *p) {
    return *(int *)p;
}

// Load and relocate. Returns 1 and fills img on success.
//
// Every segment is copied out of the read buffer into its own exact
// allocation, because the buffer is freed before returning: the
// image's code and data become ordinary host memory that the
// interpreter runs directly out of.
int c4r_load(char *path, struct c4r_image *img) {
    char *buf, *p, *data;
    int *code, *cons, *des;
    int fd, n, total, i;
    int entry, codelen, datalen, patchlen, symlen, conslen, deslen, memsz;
    int ptype, paddr, pvalu;

    img->code = 0; img->data = 0; img->entry = 0;
    img->cons = 0; img->ncons = 0; img->des = 0; img->ndes = 0;

    if (!(buf = (char *)malloc(C4R_BUF_MAX))) {
        printf("c4mp: out of memory reading %s\n", path);
        return 0;
    }
    if ((fd = open(path, 0)) < 0) {
        printf("c4mp: cannot open %s\n", path);
        free(buf);
        return 0;
    }
    total = 0;
    while ((n = read(fd, buf + total, 65536)) > 0) total = total + n;
    close(fd);
    if (total < 13) {
        printf("c4mp: %s is too short to be a .c4r\n", path);
        free(buf);
        return 0;
    }

    p = buf;
    if (!(p[0] == 'C' && p[1] == '4' && p[2] == 'R')) {
        printf("c4mp: bad signature in %s\n", path);
        free(buf);
        return 0;
    }
    // The word size is baked into every offset in the file, so a
    // mismatch is unrecoverable rather than merely inconvenient.
    if (p[4] / 8 != sizeof(int)) {
        printf("c4mp: %d-bit image, host is %d-bit\n", p[4], sizeof(int) * 8);
        free(buf);
        return 0;
    }
    // The padding word (byte offset 5) is the data MEMSZ in format v3:
    // the total in-memory data size, whose excess over datalen is
    // zero-filled BSS. v2 has no such field (memsz = datalen).
    memsz = (p[3] >= 3) ? loader_word(p + 5) : 0;
    p = p + 13;   // signature(3), version(1), wordbits(1), padding(8)

    entry    = loader_word(p); p = p + 8;
    codelen  = loader_word(p); p = p + 8;
    datalen  = loader_word(p); p = p + 8;
    patchlen = loader_word(p); p = p + 8;
    symlen   = loader_word(p); p = p + 8;
    conslen  = loader_word(p); p = p + 8;
    deslen   = loader_word(p); p = p + 8;
    if (memsz < datalen) memsz = datalen;

    // code
    p = p + 8;   // 'C' marker word
    if (!(code = (int *)malloc(codelen * 8))) { free(buf); return 0; }
    memcpy(code, p, codelen * 8);
    p = p + codelen * 8;

    // data: allocate the full in-memory size (memsz >= datalen), plus
    // one word of zero padding past the end because a guest may address
    // the last word by pointer arithmetic. Everything is zeroed, so the
    // BSS tail [datalen, memsz) is ready; only datalen bytes are copied.
    p = p + 8;   // 'D' marker
    if (!(data = (char *)malloc(memsz + 8))) { free(code); free(buf); return 0; }
    memset(data, 0, memsz + 8);
    memcpy(data, p, datalen);
    p = p + datalen;

    // patches: turn file offsets into host addresses
    p = p + 8;   // 'P' marker
    i = 0;
    while (i < patchlen) {
        ptype = loader_word(p);
        paddr = loader_word(p + 8);
        pvalu = loader_word(p + 16);
        p = p + 24;
        if (ptype == -1)      code[paddr] = (int)(code + pvalu);
        else if (ptype == -2) code[paddr] = (int)(data + pvalu);
        else if (ptype == -3) *(int *)(data + paddr) = (int)(code + pvalu);
        else if (ptype == -4) *(int *)(data + paddr) = (int)(data + pvalu);
        else {
            // A positive type is an unresolved external symbol. That
            // is the linker's job, not ours: c4rlink should have
            // consumed it, so its survival means an incomplete image.
            printf("c4mp: %s has an unresolved symbol (patch type %d)\n", path, ptype);
            free(code); free(data); free(buf);
            return 0;
        }
        ++i;
    }

    // Constructor and destructor lists. On disk each entry is a single
    // word -- the priority field the format documents is not written
    // (load-c4r.c reads only the value, with the priority read commented
    // out). Both lists point into buf, so they must be copied out.
    p = p + 8;   // 'c' marker
    cons = (int *)p;
    p = p + conslen * 8;
    des = (int *)p + 1;   // past the 'd' marker
    p = p + 8;   // 'd' marker
    p = p + deslen * 8;
    // 'S' marker and the symbol section follow. c4mp has no use for
    // symbols yet; STRC will want them when it learns to print names.

    if (conslen) {
        if (!(img->cons = (int *)malloc(conslen * 8))) {
            free(code); free(data); free(buf);
            return 0;
        }
        for (i = 0; i < conslen; ++i) img->cons[i] = (int)(code + cons[i]);
    }
    if (deslen) {
        if (!(img->des = (int *)malloc(deslen * 8))) {
            if (img->cons) free(img->cons);
            free(code); free(data); free(buf);
            return 0;
        }
        for (i = 0; i < deslen; ++i) img->des[i] = (int)(code + des[i]);
    }

    img->code  = code;
    img->data  = data;
    img->entry = code + entry;
    img->ncons = conslen;
    img->ndes  = deslen;
    free(buf);
    return 1;
}

void c4r_free(struct c4r_image *img) {
    if (img->cons) { free(img->cons); img->cons = 0; }
    if (img->des)  { free(img->des);  img->des  = 0; }
    if (img->code) { free(img->code); img->code = 0; }
    if (img->data) { free(img->data); img->data = 0; }
}
